File: frontend/src/types/prediction.ts
import { ExplanationResponse } from './explanation';
import { RecommendationItem } from './recommendation';

export interface PredictionValue {
  yield: number;
  unit: string;
}

export interface ContextPayload {
  crop: string;
  historical_mean?: number;
  historical_median?: number;
  historical_min?: number;
  historical_max?: number;
  percentile_rank?: number;
  comparison_label?: string;
}

export interface RiskPayload {
  level: 'LOW' | 'MODERATE' | 'HIGH';
  score?: number;
  drivers: string[];
  protective_factors: string[];
  component_scores?: {
    yield_deficit?: number;
    unfavorable_factors?: number;
    uncertainty_penalty?: number;
    data_quality_penalty?: number;
  };
}

export interface UncertaintyPayload {
  classification: 'LOW' | 'MODERATE' | 'HIGH';
  std_yield?: number;
  lower_bound?: number;
  upper_bound?: number;
  relative_uncertainty?: number;
  n_trees: number;
}

export interface PredictionResponse {
  prediction: PredictionValue;
  context?: ContextPayload;
  risk: RiskPayload;
  uncertainty: UncertaintyPayload;
  explanation: ExplanationResponse;
  recommendations: RecommendationItem[];
  data_quality?: {
    warnings?: string[];
    out_of_distribution?: boolean;
    extrapolation_warning?: boolean;
  };
  insights?: {
    summary?: string;
    key_drivers?: string[];
    risk_summary?: string;
  };
  metadata?: {
    project: string;
    model_name: string;
    model_version: string;
    target: string;
    target_unit: string;
  };
}

File: backend/app/schemas/prediction.py
"""
CropIQ Prediction Pydantic Schemas
Defines structured responses for the unified /predict endpoint.
"""

from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class PredictionValue(BaseModel):
    yield_: float = Field(..., alias="yield", description="Point prediction for crop yield")
    unit: str = Field(default="unconfirmed", description="Unit of yield as validated in Phase 1/2")


class ContextPayload(BaseModel):
    crop: str = Field(...)
    historical_mean: Optional[float] = None
    historical_median: Optional[float] = None
    historical_min: Optional[float] = None
    historical_max: Optional[float] = None
    percentile_rank: Optional[float] = None
    comparison_label: Optional[str] = None


class RiskPayload(BaseModel):
    level: str = Field(..., example="MODERATE")
    score: Optional[int] = Field(None, example=45)
    drivers: List[str] = Field(default_factory=list)
    protective_factors: List[str] = Field(default_factory=list)
    component_scores: Optional[Dict[str, Any]] = None


class UncertaintyPayload(BaseModel):
    classification: str = Field(..., example="LOW")
    std_yield: Optional[float] = None
    lower_bound: Optional[float] = None
    upper_bound: Optional[float] = None
    relative_uncertainty: Optional[float] = None
    n_trees: int = Field(default=300)


class PredictionResponse(BaseModel):
    prediction: PredictionValue
    context: Optional[ContextPayload] = None
    risk: RiskPayload
    uncertainty: UncertaintyPayload
    explanation: Dict[str, Any]
    recommendations: List[Dict[str, Any]] = Field(default_factory=list)
    data_quality: Dict[str, Any] = Field(default_factory=dict)
    insights: Optional[Dict[str, Any]] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)

    class Config:
        populate_by_name = True
